shared/nav.desktop.js

```diff
@@ -63,47 +55,57 @@ class Nav extends Component {
     // Restartup when we connect online.
     // If you startup while offline, you'll stay in an errored state
     window.addEventListener('online', this.props.startup)
+
+    this._renderedActiveTab = null // the last tab we actually drew
   }
 
   _handleTabsChange (e, key, payload) {
     this.props.switchTab(e)
   }
 
-  componentWillReceiveProps (nextProps) {
-    const activeTab = this.props.tabbedRouter.get('activeTab')
-    const nextActiveTab = nextProps.tabbedRouter.get('activeTab')
+  _resizeWindow () {
+    this._askedResize = false
+    const currentWindow = remote.getCurrentWindow()
 
-    // Transistioning into the login tab
-    if (activeTab !== loginTab && nextActiveTab === loginTab) {
-      this.window = remote.getCurrentWindow()
-      const [width, height] = this.window.getSize()
-      this.originalSize = {width, height}
-
-      this.window && this.window.setContentSize(globalResizing.login.width, globalResizing.login.height, true)
-      this.window && this.window.setResizable(false)
+    if (!currentWindow) {
+      return
     }
 
-    // Transistioning out of the login tab
-    if (activeTab === loginTab && nextActiveTab !== loginTab) {
-      if (this.originalSize) {
-        const {width, height} = this.originalSize
-        this.window && this.window.setSize(width, height, true)
+    if (this._renderedActiveTab === loginTab) {
+      const [width, height] = currentWindow.getSize()
+
+      // Did we actually change the size
+      if (width !== globalResizing.login.width && height !== globalResizing.login.height) {
+        this._originalSize = {width, height}
+      }
+
+      currentWindow.setContentSize(globalResizing.login.width, globalResizing.login.height, true)
+      currentWindow.setResizable(false)
+    } else {
+      if (this._originalSize) {
+        const {width, height} = this._originalSize
+        currentWindow.setSize(width, height, true)
+      } else {
+        currentWindow.setContentSize(globalResizing.normal.width, globalResizing.normal.height, true)
       }
-      this.window && this.window.setResizable(true)
+      currentWindow.setResizable(true)
     }
   }
 
+  shouldComponentUpdate (nextProps, nextState) {
+    return (nextProps.tabbedRouter.get('activeTab') !== this.props.tabbedRouter.get('activeTab'))
+  }
+
   render () {
     const activeTab = this.props.tabbedRouter.get('activeTab')
 
-    if (this.props.config.navState === Constants.navStartingUp) {
-      return (
-        <div style={{display: 'flex', flexDirection: 'column', flex: 1, alignItems: 'center', justifyContent: 'center'}}>
-          <h1>Loading...</h1>
-        </div>
-      )
+    if (!this._askedResize && (activeTab === loginTab || this._renderedActiveTab === loginTab)) {
+      this._askedResize = true
+      setImmediate(() => this._resizeWindow())
```

Comment:
Shouldn't this be in componentDidUpdate or something? render should be a pure function.